### DISCUSSION TITLE
Delphi 12 Athens: TJvTreeView fails with Access Violation after Windows RDP (mstsc.exe) remote session disconnect and connect again.

### DISCUSSION THREAD

**Discussion by alan008** [project-jedi/jvcl#176] [General]
One form with TJvTreeView
Run application in remote session, close remote session window without closing the application, connect to remote session again. JvTreeView receives WM_NOTIFY (CN_NOTIFY) and Vcl.ComCtrls CNNotify procedure call fails with AV. Example project attached.
Problem not happen with Delphi's default TTreeView, so only TJvTreeView is affected.

[JvTreeView_Delphi12_Test.zip](https://github.com/project-jedi/jvcl/files/13900999/JvTreeView_Delphi12_Test.zip)

**Comment by alan008** [comment]
TJvCheckTreeView is also affected by this problem

**Comment by alan008** [comment]
Looks like JvExComCtrls is "somehow incompatible" with current Vcl.ComCtrls.pas implementation in Delphi 12.

**Comment by alan008** [comment]
Looks like same as https://github.com/project-jedi/jvcl/discussions/173

The stack trace is not very informative, because JvExComCtrls.pas (more exactly -JvExControls.macros) hacks current Delphi 12 implementation probably incorrectly.

**Reply by obones** [reply]
Not very informative to you, but I'd really like to have it anyway

**Reply by alan008** [reply]
:775af932 KERNELBASE.RaiseException + 0x62
:009a58a9 TJvTreeView.CNNotify + $269
:009a58a9 TJvTreeView.CNNotify + $269
Vcl.Controls.TWinControl.WndProc((48206, 1715608, 5498496, 0, 11672, 26, (), 59008, 83, (), 0, 0, ()))
Vcl.ComCtrls.TCustomTreeView.WndProc((48206, 1715608, 5498496, 0, 11672, 26, (), 59008, 83, (), 0, 0, ()))
:008f53b3 TCustomTreeView.WndProc + $87
Vcl.Controls.TControl.Perform(???,???,5498496)
Vcl.Controls.DoControlMsg(???,???)
Vcl.Controls.TWinControl.WMNotify((78, (), 1715608, $53E680, 0))
Vcl.Controls.TControl.WndProc((78, 1715608, 5498496, 0, 11672, 26, (), 59008, 83, (), 0, 0, ()))
Vcl.Controls.TWinControl.WndProc((78, 1715608, 5498496, 0, 11672, 26, (), 59008, 83, (), 0, 0, ()))
Vcl.Forms.TCustomForm.WndProc((78, 1715608, 5498496, 0, 11672, 26, (), 59008, 83, (), 0, 0, ()))
Vcl.Controls.TWinControl.MainWndProc(???)
System.Classes.StdWndProc(12911884,78,1715608,5498496)
:75f511ab USER32.AddClipboardFormatListener + 0x4b
:75f4816a ; C:\Windows\SysWOW64\USER32.dll
:75f47539 ; C:\Windows\SysWOW64\USER32.dll
:75f45b7f USER32.SendMessageW + 0x6f
:746b9887 ; C:\Windows\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\comctl32.dll
:746798a7 ; C:\Windows\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\comctl32.dll
:7467969a ; C:\Windows\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\comctl32.dll
:7468fc6d ; C:\Windows\WinSxS\x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.19041.3636_none_a863d714867441db\comctl32.dll
:75f511ab USER32.AddClipboardFormatListener + 0x4b
:75f4816a ; C:\Windows\SysWOW64\USER32.dll
:75f47837 ; C:\Windows\SysWOW64\USER32.dll
:75f459db USER32.CallWindowProcW + 0x1b
Vcl.Controls.TWinControl.DefaultHandler(???)
:008c83fa TWinControl.DefaultHandler + $136
:008c82a4 TWinControl.WndProc + $6AC
:008f53b3 TCustomTreeView.WndProc + $87
:008c7767 TWinControl.MainWndProc + $2F
:0084b7b6 StdWndProc + $16
:75f511ab USER32.AddClipboardFormatListener + 0x4b
:75f4816a ; C:\Windows\SysWOW64\USER32.dll
:75f47539 ; C:\Windows\SysWOW64\USER32.dll
:75f45b7f USER32.SendMessageW + 0x6f
:008825bc TreeView_SelectItem + $14
:009a4dc7 TJvTreeView.CreateWnd + $13
:008c759b TWinControl.UpdateControlState + $C3
:008c71b0 TWinControl.CMRecreateWnd + $44
:008c2c5d TControl.WndProc + $2C1
:008c82a4 TWinControl.WndProc + $6AC
:008f53b3 TCustomTreeView.WndProc + $87
:008c2893 TControl.Perform + $27
:008c716b TWinControl.RecreateWnd + $17
:008c82a4 TWinControl.WndProc + $6AC
:008f53b3 TCustomTreeView.WndProc + $87
:008c6664 TWinControl.Broadcast + $30


![2024-01-12_12-58-51](https://github.com/project-jedi/jvcl/assets/2262134/7c558ada-d3f2-499d-9d2d-590f9d409d17)

**Reply by alan008** [reply]
The real place and error itself differs depending on Memory manager, Project and so on. Debugger shows not real lines. In this one-form test projects it appears as "invalid class typecast", but in real life project it was c0000005 and project full crash.

**Reply by alan008** [reply]
Found simple way to reproduce this error without remote session, just enable visual styles and try to do TreeView.StyleName := 'Some Style'. Example attached, try to click Button1 at runtime

[JvTreeView_Delphi12_Test.zip](https://github.com/project-jedi/jvcl/files/13917857/JvTreeView_Delphi12_Test.zip)

**Reply by alan008** [reply]
Any news on this? At least reproducible?

**Reply by obones** [reply]
> Any news on this? At least reproducible?

You do realize that we are a very small group of volunteers working on this project right?
We can only use our spare time for this and it has to be shared among many other things.

Having a test application helped quite a lot here, and as I suspected this has nothing to do with `JvExComCtrls` but all to do with a change inside `TTreeNodes.ReadNodeData` from the Delphi 12 VCL itself.

Anyway, a fix is already in the repository.

**Reply by alan008** [reply]
Great thanks!!! I'll test and write if it helps!
I'm working with Delphi since 2000s, and of couse I understand, that Delphi's community is still holding on individuals great personal effort (more often - voluntary). Anyway, bugs and compatibility with latest/future Delphi releases is a necessary requirement, because it allows to drive innovations in (a slightly vanishing) Delphi world. When large project migrates to new Delphi version, it's always a bit of pain, because (in this particular case), I cannot replace quickly TJvTreeView in 100+ forms into some other component.

**Reply by alan008** [reply]
Tested OK, the problem is solved! Thanks again!
